File: MergekSortedLists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        ListNode* head = new ListNode(INT_MIN + 2);
        ListNode* thead = head;
        priority_queue<pair<int, ListNode*>> pq;
        int k = lists.size();
        if(k == 0)return nullptr;
        for(int i = 0;i < k;i++){
            ListNode* tl = lists[i];
            if(tl != nullptr)pq.push(make_pair(-1 * (tl->val), tl));
        }
        
        while(!pq.empty()){
            pair<int, ListNode*> p = pq.top();
            ListNode* s = p.second;
            thead -> next = s;
            thead = s;
            pq.pop();
            
            if(s -> next != nullptr)pq.push(make_pair(-1 * (s -> next -> val), s -> next));
        }
        
        // if(thead -> val == INT_MIN)return nullptr;
        
        thead -> next = nullptr;
        return head -> next;
        
    }
};

File: MergekSortedLists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class CustomComparator implements Comparator<ListNode>{

    public int compare(ListNode a, ListNode b){
        return a.val - b.val;
    }
} 
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        
        PriorityQueue<ListNode> minHeap = new PriorityQueue<ListNode>(new CustomComparator());
        int n = lists.length;
        ListNode h = new ListNode(-1);
        ListNode hp = h;
        for(int i = 0;i < n;i++)if(lists[i] != null)minHeap.add(lists[i]);

        while(minHeap.size() != 0){
            ListNode e = minHeap.poll();h.next = e;
            h = e;
            if(e.next != null)minHeap.add(e.next);
        }
        return hp.next;
    }
}
